sklearn/linear_model/ridge.py

```diff
@@ -299,8 +311,9 @@ def ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
         if np.atleast_1d(sample_weight).ndim > 1:
             raise ValueError("Sample weights must be 1D array or scalar")
 
-        # Sample weight can be implemented via a simple rescaling.
-        X, y = _rescale_data(X, y, sample_weight)
+        if solver == 'cholesky':
+            # Sample weight can be implemented via a simple rescaling.
+            X, y = _rescale_data(X, y, sample_weight)
```

Comment:
I misunderstood the doc:
> If sample_weight is set, then the solver will automatically be set to 'cholesky'

I guess it means: 
>If sample_weight is set and solver='auto', the solver will be set to 'cholesky'

which corresponds to actual behavior.

In the sparse case with `fit_intercept=True`, we cannot center the data, so the rescaling is not correct. As 'sag' solver handles correctly the sample weights, I did not rescale in this case.

I guess I can change it to `if solver != 'sag'`, to keep previous behavior for other solvers.
Previous behavior looks however suspicious to me in the sparse case with intercept (see issue #4710).